### PR TITLE
Add ImpersonateEvents for SnappyMail (and others)

### PR DESCRIPTION
**Pull request by the-djmaze**
More info at #179

### CODE REVIEW

**Comment by the-djmaze** [comment]
> Does it carry enough information to know which user is impersonated

For me personally i don't care which user. Just "impersonate" is enough for me to force logout.

I've just extended it a bit if other apps also want to use it.
Yours would even be better for other apps that might want to know details regarding the impersonating.

Just do what you think is right.

**Comment by blizzz** [comment]
@the-djmaze added a commit, does this work for you? I removed the legacy emitter, it should not be used anymore. It carries only some legacy bits (which should be phased out…)

**Comment by blizzz** [comment]
@the-djmaze does it work for you? ^

**Comment by the-djmaze** [comment]
I will check!

File: lib/Controller/LogoutController.php [hunk 1]
```diff
@@ -60,7 +66,10 @@ public function logout(string $userId): JSONResponse {
 			);
 		}
 
-		$this->userSession->setUser($user);
+		$impersonatee = $this->userManager->get($userId);
```
Comment (nickvergessen) [inline lib/Controller/LogoutController.php:69]: I love the more clearer distinction on the name of the Getter function on the event.
...tee vs ...tor is visually so similar. `$impersonatedUser` would be more different.

**Review by nickvergessen — COMMENTED** [review]
Fine in general
Just two suggestions

File: lib/Controller/LogoutController.php [hunk 2]
```diff
@@ -34,23 +38,25 @@ public function __construct($appName,
...
 
 	/**
 	 * @UseSession
 	 * @NoAdminRequired
 	 */
 	public function logout(string $userId): JSONResponse {
-		$user = $this->session->get('oldUserId');
-		$user = $this->userManager->get($user);
+		$impersonator = $this->session->get('oldUserId');
+		$impersonator = $this->userManager->get($impersonator);
```
[the remark this answers is not in the document]
Comment (blizzz) [reply lib/Controller/LogoutController.php:57]: Had the same feeling and almost changed it back then. Well, almost. :+1:

File: lib/Controller/LogoutController.php (hunk 1 above)
Comment (blizzz) [reply lib/Controller/LogoutController.php:69]: Same gut feeling, :+1:

Comment (blizzz) [reply lib/Controller/LogoutController.php:69]: Ah, actually as i did in the Event class… =D

**Comment by the-djmaze** [comment]
On logout i get:
`Argument 2 passed to OCA\\Impersonate\\Events\\AImpersonateEvent::__construct() must implement interface OCP\\IUser, null given`

**Comment by blizzz** [comment]
> On logout i get: `Argument 2 passed to OCA\\Impersonate\\Events\\AImpersonateEvent::__construct() must implement interface OCP\\IUser, null given`
> 
> Further it works as expected and helps a lot!

I'll have another look, thanks for spotting it!

**Comment by blizzz** [comment]
> > On logout i get: `Argument 2 passed to OCA\\Impersonate\\Events\\AImpersonateEvent::__construct() must implement interface OCP\\IUser, null given`
> > Further it works as expected and helps a lot!
> 
> I'll have another look, thanks for spotting it!

Probably needs adjustements on the impersonate_logout.js code, perhaps getting rid of jquery traces :see_no_evil: needs more looking into

**Comment by blizzz** [comment]
Latest commit fixes logout

**Review by come-nc — APPROVED** [review]

**Pull request merged**